web/src/main/webapp/heatMapArray/view.ts

```diff
@@ -0,0 +1,288 @@
+import {RemoteObjectView, CallbackCollector} from "../rpc";
+import {Dialog} from "../dialog";
+import {RangeInfo, TableView, Schema, IColumnDescription, isNumeric, isCategorical, DistinctStrings} from "../table";
+import {BasicColStats} from "../histogramBase";
+import {FullPage, Size, Resolution, IHtmlElement, ScrollBar, IScrollTarget} from "../ui";
+import {Pair, Triple} from "../util";
+import {ColumnAndRange} from "../histogramBase";
+import {HeatMapData} from "../heatmap";
+import {HeatMapArrayArgs, HeatMapArrayData} from "./types";
+import d3 = require('d3');
+
+class ColorMap {
+    public logScale: boolean;
+    constructor(public max) {
+        this.logScale = max > 20;
+    }
+
+    public setLogScale(logScale: boolean) {
+        this.logScale = logScale;
+    }
+
+    public apply(x) {
+        if (this.logScale)
+            return this.applyLog(x);
+        else
+            return this.applyLinear(x);
+    }
+
+    private applyLinear(x) {
+        return this.map(x / this.max);
+    }
+
+    private applyLog(x) {
+        return this.map(Math.log(x) / Math.log(this.max));
+    }
+
+    private map(x: number) {
+        return d3.interpolateWarm(x);
+    }
+}
+
+class CompactHeatMapView implements IHtmlElement {
+    // Size of a single heatmap in pixels
+    public static readonly size: Size = {
+        width: 200,
+        height: 200
+    }
+    private static maxTextLabelLength = 10;
+
+    private topLevel: HTMLElement;
+    private chart: any;
+
+    constructor(private binLabel: string) {
+        this.topLevel = document.createElement("div")
+        if (binLabel.length > CompactHeatMapView.maxTextLabelLength) {
+            binLabel = binLabel.slice(0, CompactHeatMapView.maxTextLabelLength) + "...";
+        }
+        this.topLevel.appendChild(document.createElement("p")).textContent = binLabel;
+        this.chart = d3.select(this.topLevel).append("svg")
+            .attr("width", CompactHeatMapView.size.width)
+            .attr("height", CompactHeatMapView.size.height)
+    }
+
+    public put(x, y, val) {
+        this.chart.append("rect")
+            .attr("x", x * Resolution.minDotSize)
```

Comment:
You cannot hardwire the dotsize, it has to be computed as the width divided by the number of buckets.
This changes when you resize the window.